### PR TITLE
fix(derive_code_mappings): Ignore file paths without an extension

### PR DESCRIPTION
**Pull request by armenzg**
This fixes WOR-2478

### CODE REVIEW

File: src/sentry/integrations/utils/code_mapping.py [hunk 1]
```diff
@@ -33,10 +33,14 @@ class CodeMapping(NamedTuple):
...
 
+class UnsupportedFrameFilename(Exception):
+    pass
+
+
 def get_extension(file_path: str) -> str:
     extension = ""
     if file_path:
-        ext_period = file_path.find(".")
+        ext_period = file_path.rfind(".")
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:43]: This is a fix for `"./app/utils/handleXhrErrorResponse.tsx`

File: src/sentry/integrations/utils/code_mapping.py [hunk 2]
```diff
@@ -73,8 +77,13 @@ def filter_source_code_files(files: List[str]) -> List[str]:
 # XXX: Look at sentry.interfaces.stacktrace and maybe use that
 class FrameFilename:
     def __init__(self, frame_file_path: str) -> None:
+        # XXX: Using regexes would be better
+        if not frame_file_path or frame_file_path.find("<") > -1:
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:81]: This is to handle things like `<frozen importlib._bootstrap>`

File: src/sentry/integrations/utils/code_mapping.py [hunk 3]
```diff
@@ -73,8 +77,13 @@ def filter_source_code_files(files: List[str]) -> List[str]:
...
+            raise UnsupportedFrameFilename
         self.full_path = frame_file_path
         self.extension = get_extension(frame_file_path)
+        if not self.extension:
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:85]: This handles `O$t` or `/foo/bar/baz`.

File: src/sentry/integrations/utils/code_mapping.py [hunk 4]
```diff
@@ -120,29 +129,35 @@ def __eq__(self, other) -> bool:  # type: ignore
         return self.full_path == other.full_path  # type: ignore
 
 
+def stacktrace_buckets(stacktraces: List[str]) -> Dict[str, List[FrameFilename]]:
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:132]: This has been extracted outside of the class below.

File: src/sentry/integrations/utils/code_mapping.py [hunk 5]
```diff
@@ -120,29 +129,35 @@ def __eq__(self, other) -> bool:  # type: ignore
...
+        try:
+            frame_filename = FrameFilename(stacktrace_frame_file_path)
+            # Any files without a top directory will be grouped together
+            bucket_key = frame_filename.root if frame_filename.root else NO_TOP_DIR
+
+            if not buckets.get(bucket_key):
+                buckets[bucket_key] = []
+            buckets[bucket_key].append(frame_filename)
+
+        except UnsupportedFrameFilename:
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:144]: This except block is new.

File: src/sentry/integrations/utils/code_mapping.py [hunk 6]
```diff
@@ -120,29 +129,35 @@ def __eq__(self, other) -> bool:  # type: ignore
...
+    return buckets
+
+
 # call generate_code_mappings() after you initialize CodeMappingTreesHelper
 class CodeMappingTreesHelper:
     def __init__(self, trees: Dict[str, RepoTree]):
         self.trees = trees
         self.code_mappings: Dict[str, CodeMapping] = {}
 
-    def stacktrace_buckets(self, stacktraces: List[str]) -> Dict[str, List[FrameFilename]]:
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:129]: Moved out.

File: tests/sentry/integrations/utils/test_code_mapping.py [hunk 7]
```diff
@@ -44,11 +47,63 @@ def test_should_not_include(self):
         ]:
             assert should_include(file) is False
 
-    def test_get_extension(self):
-        assert get_extension("") == ""
-        assert get_extension(None) == ""
-        assert get_extension("f.py") == "py"
-        assert get_extension("f.xx") == "xx"
+
+def test_get_extension():
```
Comment (armenzg) [inline tests/sentry/integrations/utils/test_code_mapping.py:51]: I'm moving this out as it's not related to files included in a repo.

File: tests/sentry/integrations/utils/test_code_mapping.py [hunk 8]
```diff
@@ -44,11 +47,63 @@ def test_should_not_include(self):
...
+    assert get_extension("") == ""
+    assert get_extension(None) == ""
+    assert get_extension("f.py") == "py"
+    assert get_extension("f.xx") == "xx"
+    assert get_extension("./app/utils/handleXhrErrorResponse.tsx") == "tsx"
+    assert get_extension("[native code]") == ""
+    assert get_extension("/foo/bar/baz") == ""
+    assert get_extension("/gtm.js") == "js"
```
Comment (armenzg) [inline tests/sentry/integrations/utils/test_code_mapping.py:59]: There's 4 new cases.

File: tests/sentry/integrations/utils/test_code_mapping.py [hunk 9]
```diff
@@ -44,11 +47,63 @@ def test_should_not_include(self):
...
+
+
+def test_buckets_logic():
+    stacktraces = [
+        "<anonymous>",  # Garbage
+        "<frozen importlib._bootstrap>",
+        "[native code]",
+        "O$t",
+        "/foo/bar/baz",  # no extension
```
Comment (armenzg) [inline tests/sentry/integrations/utils/test_code_mapping.py:68]: The ones above are new cases.

File: tests/sentry/integrations/utils/test_code_mapping.py [hunk 10]
```diff
@@ -44,11 +47,63 @@ def test_should_not_include(self):
...
+    assert get_extension(None) == ""
+    assert get_extension("f.py") == "py"
+    assert get_extension("f.xx") == "xx"
+    assert get_extension("./app/utils/handleXhrErrorResponse.tsx") == "tsx"
+    assert get_extension("[native code]") == ""
+    assert get_extension("/foo/bar/baz") == ""
+    assert get_extension("/gtm.js") == "js"
+
+
+def test_buckets_logic():
```
Comment (armenzg) [inline tests/sentry/integrations/utils/test_code_mapping.py:62]: Moved outside of the class.

File: tests/sentry/integrations/utils/test_code_mapping.py [hunk 11]
```diff
@@ -44,11 +47,63 @@ def test_should_not_include(self):
...
+        "app://foo.js",
+        "./app/utils/handleXhrErrorResponse.tsx",
+        "getsentry/billing/tax/manager.py",
+        "/gtm.js",  # Top source; starts with backslash
+        "ssl.py",
+    ]
+    buckets = stacktrace_buckets(stacktraces)
+    assert buckets == {
+        "./app": [FrameFilename("./app/utils/handleXhrErrorResponse.tsx")],
+        "NO_TOP_DIR": [FrameFilename("/gtm.js"), FrameFilename("ssl.py")],
```
Comment (armenzg) [inline tests/sentry/integrations/utils/test_code_mapping.py:80]: `/gtm.js` in here is a new case.

File: tests/sentry/integrations/utils/test_code_mapping.py [hunk 12]
```diff
@@ -44,11 +47,63 @@ def test_should_not_include(self):
...
+        "app:": [FrameFilename("app://foo.js")],
+        "getsentry": [FrameFilename("getsentry/billing/tax/manager.py")],
+    }
+
+
+class TestFrameFilename(TestCase):
```
Comment (armenzg) [inline tests/sentry/integrations/utils/test_code_mapping.py:86]: Move these tests into their our group. No changes in here.

File: tests/sentry/integrations/utils/test_code_mapping.py [hunk 13]
```diff
@@ -76,42 +131,6 @@ def setUp(self):
...
-
-    def test_frame_filename_no_package(self):
-        ff = FrameFilename("foo.py")
-        assert ff.root == ""
-        assert ff.dir_path == ""
-        assert ff.file_name == "foo.py"
-
-    def test_frame_filename_repr(self):
-        path = "getsentry/billing/tax/manager.py"
-        assert FrameFilename(path).__repr__() == f"FrameFilename: {path}"
```
Comment (armenzg) [inline tests/sentry/integrations/utils/test_code_mapping.py:98]: These four tests are now grouped together.

**Review by scttcper — APPROVED** [review]

**Pull request merged**